### ISSUE TITLE
lock_picks issue for viewing other player's picks

### ISSUE THREAD

**Issue by jbholden** [blreams/collegefootballpick10#61]
One complication I have run into is when can a player view other players picks.  It is currently setup as follows

```
if it is before the pick deadline, a player cannot view other player's picks
if it is after the pick deadline, a player can view other player's picks
if lock_picks is None (i.e. not deadline set), a player can view other player's picks
```

When the commissioner is setting up games, lock_picks will get set to a value in the past.  As a result, a player will be able to view other player's picks at this time.  This is not a problem if no picks have been made yet.

However, if any picks are made, and the commissioner chooses to lock picks, then other player's picks will become visible.  

Is it possible that lock_picks will get set during the "enter picks" phase, and thus accidentally allowing the viewing of other player's picks?

**Comment by blreams** [comment]
The more I think about this issue, the more I think we need to have separate fields for the pick deadline and for locking picks. You mentioned this as an option in Issue #54. Let me walk through how this might work and see where it takes me.

Suppose we have two fields: 1) lock_picks (boolean), 2) pick_deadline (datetime). When a commissioner first creates a set of picks it won't matter if there is a pick deadline as long as lock_picks == True. Once the pick setup is finalized, the lock_picks field can be set to False (but only if a pick_deadline has been set). 

The editweek URL can still use Week.lock_picks as the semaphore to prevent the race condition between a commissioner editing a week's games vs. a player entering picks for that week.

Your code can use pick_deadline instead of lock_picks to determine pool state and whether or not to allow a player to view other player's picks.

**Comment by jbholden** [comment]
I agree that this is a good idea.

**Comment by blreams** [comment]
I created a branch (weekmodelchange) and I started coding changes to the Week model so it has separate locks_picks and pick_deadline fields. I made many of the required changes, but I'm stuck on some of the unit test failures. 

For example:

```
$ python manage.py test pick10.tests.test_calculator.CalculatorTests.test_t36_3_get_player_submit_time
Creating test database for alias 'default'...
E
======================================================================
ERROR: test_t36_3_get_player_submit_time (pick10.tests.test_calculator.CalculatorTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/blreams/Documents/django/collegefootballpick10/pick10/tests/test_calculator.py", line 622, in test_t36_3_get_player_submit_time
    self.__t36_week_lock_picks_not_specified()
  File "/home/blreams/Documents/django/collegefootballpick10/pick10/tests/test_calculator.py", line 2618, in __t36_week_lock_picks_not_specified
    self.assertIsNone(self.calc.get_player_submit_time(player,week))
  File "/home/blreams/Documents/django/collegefootballpick10/pick10/calculator.py", line 411, in get_player_submit_time
    if self.__submit_time_invalid(week,latest_time):
  File "/home/blreams/Documents/django/collegefootballpick10/pick10/calculator.py", line 438, in __submit_time_invalid
    picks_entered_after_pick_deadline = submit_time > week.pick_deadline
TypeError: can't compare offset-naive and offset-aware datetimes

----------------------------------------------------------------------
Ran 1 test in 6.677s

FAILED (errors=1)
Destroying test database for alias 'default'...
```

It's obviously not happy comparing the two datatime objects (one naive and the other is TZ aware). What I can't figure out is why submit_time is naive. It looks to me like it comes from the DB Pick.updated (which should be TZ aware).

**Comment by blreams** [comment]
OK, I think I got to the bottom of it. I didn't realize that inside the tests, the values for Pick.created and Pick.updated were being overridden. And, really, the problem was that those tests needed to take into account having both Week.lock_picks and Week.pick_deadline as separate fields. So, I've got the calculator unit tests passing again. I still need to run through all the other unit tests.

Then there are the functional tests. I think a few changes are warranted there, but I haven't looked into what those are yet.

**Comment by blreams** [comment]
At this point I think the weekmodelchange branch is ready to be merged to master. But I don't want to do this until you have had a chance to review the changes and maybe run additional tests. I've run all the unit tests that I know about. I've also run some of the functional tests, but for some reason I'm getting long hangs for some tests in test_enter_picks.py.

I created a pull request (Issue #65) for you to take a look.

**Comment by jbholden** [comment]
I was seeing some long hangs when running tests at one time due to the delay in loading the jquery or bootstrap script over the web, but my web connection was really slow.

I know this was probably fixed when we switched to static loading of these scripts.

When say long hangs, do the hangs occur with the browsers window open or with it closed?  (assuming you are using firefox browser).  A couple of the tests take a little longer because it tries every possible pick and ensures it gets set correctly in the database.  I believe the window will stay open though.

**Comment by blreams** [comment]
I thought I had pretty much narrowed it down to functional_tests.test_enter_picks.EnterPicksTest.test_missing_one_pick. I know that is a particularly long test...but when I added a set_trace() and stepped through it, it ran to completion and passed. I also tried skipping that particular test, but it seemed to still hang. I'm wondering if there is some dependency between tests that isn't obvious.

BTW, I'm using phantomjs, so it isn't obvious what the UI looks like at the time of the hang.
